benchmarks/union_find/union_find.cpp

```diff
@@ -0,0 +1,167 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <ArborX_DetailsSortUtils.hpp>
+#include <ArborX_DetailsUnionFind.hpp>
+#include <ArborX_DetailsUtils.hpp>
+#include <ArborX_Version.hpp>
+
+#include <Kokkos_Core.hpp>
+#include <Kokkos_Random.hpp>
+#include <Kokkos_Timer.hpp>
+
+#include <boost/program_options.hpp>
+
+struct UnweightedEdge
+{
+  unsigned int source;
+  unsigned int target;
+};
+
+template <typename ExecutionSpace>
+Kokkos::View<UnweightedEdge *, typename ExecutionSpace::memory_space>
+buildEdges(ExecutionSpace const &exec_space, int num_edges, bool allow_loops)
+{
+  using MemorySpace = typename ExecutionSpace::memory_space;
+  Kokkos::View<UnweightedEdge *, MemorySpace> edges(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::Benchmark::edges"),
+      num_edges);
+
+  Kokkos::Random_XorShift1024_Pool<MemorySpace> rand_pool(1984);
+  if (allow_loops)
+  {
+    Kokkos::parallel_for(
+        "ArborX::Bechmark::init_edges",
+        Kokkos::RangePolicy<ExecutionSpace>(ExecutionSpace{}, 0, num_edges),
+        KOKKOS_LAMBDA(unsigned i) {
+          auto rand_gen = rand_pool.get_state();
+          do
+          {
+            edges(i) = {rand_gen.urand() % num_edges,
+                        rand_gen.urand() % num_edges};
+          } while (edges(i).source == edges(i).target); // no self loops
+          rand_pool.free_state(rand_gen);
```

Comment:
Short answer: I don't see how making sure there is a cycle improves anything, as we are only concerned with performance.

Long answer: actually guaranteeing a cycle for a random graph generation is non-trivial, unless one inserts it manually (which I think does not really change anything). Theoretically, if we generate $n$ unique edges for $n$ vertices, we would guarantee a cycle. However, the issue is that generation of random unique edges is hard (or at least I don't see an easy way). So generating $n$ edges would either produce a cycle, or non-unique edges (depending on the purpose, this could be considered a cycle).

Reply:
I was assuming that having a cycle changes the performance and that's why we wanted to test both cases. If it doesn't change the performance, then I agree there is no point in doing it.

Reply:
The cycle should not affect the performance much. The reason why there is a switch in the first place is that there are some algorithms (like the one from Wang's HDBSCAN paper [code](https://github.com/wangyiqiu/hdbscan/blob/e120dcfbd20283542d5bcd359cf6a1c10296058b/src/unionFind.h#L37-L88)) that only work correctly when the graph has no cycles.